codalab/apps/web/views.py

```diff
@@ -1949,6 +1949,138 @@ def get_success_url(self):
         return reverse('competitions:dumps', kwargs={'competition_pk': dump.competition.pk})
 
 
+class CompetitionWidgetsView(LoginRequiredMixin, DetailView):
+    queryset = models.Competition.objects.all()
+    template_name = 'web/competitions/widgets.html'
+
+    def get_object(self, queryset=None):
+        self.object = super(CompetitionWidgetsView, self).get_object()
+        if self.request.user != self.object.creator and self.request.user not in self.object.admins.all():
+            raise Http404()
+        return self.object
+
+
+class WidgetMixin(object):
+
+    def get_context_data(self, **kwargs):
+        context = super(WidgetMixin, self).get_context_data(**kwargs)
+        context['current_phase'] = get_current_phase(self.object)
+        return context
+
+
+class CompetitionSubmissionWidgetView(WidgetMixin, DetailView):
+    queryset = models.Competition.objects.all()
+    template_name = 'web/widget_iframes/submission.html'
+
+    def get_context_data(self, **kwargs):
+        context = super(CompetitionSubmissionWidgetView, self).get_context_data(**kwargs)
+        context['phase'] = get_current_phase(self.object)
+        competition = self.object
+
+        if settings.USE_AWS:
+            context['form'] = forms.SubmissionS3UploadForm
+
+        if competition.participants.filter(user__in=[self.request.user]).exists():
+            participant = competition.participants.get(user=self.request.user)
+            if participant.status.codename == models.ParticipantStatus.APPROVED:
+                phase = context['phase']
```

Comment:
For readability/style wouldn't it be better to instantiate the phase first, and then assign the context to the phase instead?

```
phase = get_current_phase(self.object)
context['phase'] = phase
```

Reply:
Yeah I was lazy about this, nice catch, moving it